File: utilcode/src/main/java/com/blankj/utilcode/util/StringUtils.java
package com.blankj.utilcode.util;

/**
 * <pre>
 *     author: Blankj
 *     blog  : http://blankj.com
 *     time  : 2016/08/16
 *     desc  : 字符串相关工具类
 * </pre>
 */
public final class StringUtils {

    private StringUtils() {
        throw new UnsupportedOperationException("u can't instantiate me...");
    }

    /**
     * 判断字符串是否为null或长度为0
     *
     * @param s 待校验字符串
     * @return {@code true}: 空<br> {@code false}: 不为空
     */
    public static boolean isEmpty(final CharSequence s) {
        return s == null || s.length() == 0;
    }

    /**
     * 判断字符串是否为null或全为空格
     *
     * @param s 待校验字符串
     * @return {@code true}: null或全空格<br> {@code false}: 不为null且不全空格
     */
    public static boolean isTrimEmpty(final String s) {
        return (s == null || s.trim().length() == 0);
    }

    /**
     * 判断字符串是否为null或全为空白字符
     *
     * @param s 待校验字符串
     * @return {@code true}: null或全空白字符<br> {@code false}: 不为null且不全空白字符
     */
    public static boolean isSpace(final String s) {
        if (s == null) return true;
        for (int i = 0, len = s.length(); i < len; ++i) {
            if (!Character.isWhitespace(s.charAt(i))) {
                return false;
            }
        }
        return true;
    }

    /**
     * 判断两字符串是否相等
     *
     * @param a 待校验字符串a
     * @param b 待校验字符串b
     * @return {@code true}: 相等<br>{@code false}: 不相等
     */
    public static boolean equals(final CharSequence a, final CharSequence b) {
        if (a == b) return true;
        int length;
        if (a != null && b != null && (length = a.length()) == b.length()) {
            if (a instanceof String && b instanceof String) {
                return a.equals(b);
            } else {
                for (int i = 0; i < length; i++) {
                    if (a.charAt(i) != b.charAt(i)) return false;
                }
                return true;
            }
        }
        return false;
    }

    /**
     * 判断两字符串忽略大小写是否相等
     *
     * @param a 待校验字符串a
     * @param b 待校验字符串b
     * @return {@code true}: 相等<br>{@code false}: 不相等
     */
    public static boolean equalsIgnoreCase(final String a, final String b) {
        return a == null ? b == null : a.equalsIgnoreCase(b);
    }

    /**
     * null转为长度为0的字符串
     *
     * @param s 待转字符串
     * @return s为null转为长度为0字符串，否则不改变
     */
    public static String null2Length0(final String s) {
        return s == null ? "" : s;
    }

    /**
     * 返回字符串长度
     *
     * @param s 字符串
     * @return null返回0，其他返回自身长度
     */
    public static int length(final CharSequence s) {
        return s == null ? 0 : s.length();
    }

    /**
     * 首字母大写
     *
     * @param s 待转字符串
     * @return 首字母大写字符串
     */
    public static String upperFirstLetter(final String s) {
        if (isEmpty(s) || !Character.isLowerCase(s.charAt(0))) return s;
        return String.valueOf((char) (s.charAt(0) - 32)) + s.substring(1);
    }

    /**
     * 首字母小写
     *
     * @param s 待转字符串
     * @return 首字母小写字符串
     */
    public static String lowerFirstLetter(final String s) {
        if (isEmpty(s) || !Character.isUpperCase(s.charAt(0))) return s;
        return String.valueOf((char) (s.charAt(0) + 32)) + s.substring(1);
    }

    /**
     * 反转字符串
     *
     * @param s 待反转字符串
     * @return 反转字符串
     */
    public static String reverse(final String s) {
        int len = length(s);
        if (len <= 1) return s;
        int mid = len >> 1;
        char[] chars = s.toCharArray();
        char c;
        for (int i = 0; i < mid; ++i) {
            c = chars[i];
            chars[i] = chars[len - i - 1];
            chars[len - i - 1] = c;
        }
        return new String(chars);
    }

    /**
     * 转化为半角字符
     *
     * @param s 待转字符串
     * @return 半角字符串
     */
    public static String toDBC(final String s) {
        if (isEmpty(s)) return s;
        char[] chars = s.toCharArray();
        for (int i = 0, len = chars.length; i < len; i++) {
            if (chars[i] == 12288) {
                chars[i] = ' ';
            } else if (65281 <= chars[i] && chars[i] <= 65374) {
                chars[i] = (char) (chars[i] - 65248);
            } else {
                chars[i] = chars[i];
            }
        }
        return new String(chars);
    }

    /**
     * 转化为全角字符
     *
     * @param s 待转字符串
     * @return 全角字符串
     */
    public static String toSBC(final String s) {
        if (isEmpty(s)) return s;
        char[] chars = s.toCharArray();
        for (int i = 0, len = chars.length; i < len; i++) {
            if (chars[i] == ' ') {
                chars[i] = (char) 12288;
            } else if (33 <= chars[i] && chars[i] <= 126) {
                chars[i] = (char) (chars[i] + 65248);
            } else {
                chars[i] = chars[i];
            }
        }
        return new String(chars);
    }
}

File: utilcode-kotlin/src/main/java/com/blankj/utilcode/util/StringUtils.kt
@file:JvmName("StringUtils")

package com.blankj.utilcode.util

/**
 * Return whether the string is null or 0-length.
 *
 * @param s The string.
 * @return `true`: yes<br></br> `false`: no
 */
fun isEmpty(s: CharSequence?): Boolean {
    return s == null || s.isEmpty()
}

/**
 * Return whether the string is null or whitespace.
 *
 * @param s The string.
 * @return `true`: yes<br></br> `false`: no
 */
fun isTrimEmpty(s: String?): Boolean {
    return s == null || s.trim { it <= ' ' }.isEmpty()
}

/**
 * Return whether the string is null or white space.
 *
 * @param s The string.
 * @return `true`: yes<br></br> `false`: no
 */
fun isSpace(s: String?): Boolean {
    if (s == null) return true
    var i = 0
    val len = s.length
    while (i < len) {
        if (!Character.isWhitespace(s[i])) {
            return false
        }
        ++i
    }
    return true
}

/**
 * Return whether string1 is equals to string2.
 *
 * @param s1 The first string.
 * @param s2 The second string.
 * @return `true`: yes<br></br>`false`: no
 */
fun equals(s1: CharSequence?, s2: CharSequence?): Boolean {
    if (s1 === s2) return true
    if (s1 != null && s2 != null) {
        val length: Int = s1.length
        if (length == s2.length) {
            if (s1 is String && s2 is String) {
                return s1 == s2
            } else {
                for (i in 0 until length) {
                    if (s1[i] != s2[i]) return false
                }
                return true
            }
        }
    }
    return false
}

/**
 * Return whether string1 is equals to string2, ignoring case considerations..
 *
 * @param s1 The first string.
 * @param s2 The second string.
 * @return `true`: yes<br></br>`false`: no
 */
fun equalsIgnoreCase(s1: String?, s2: String?): Boolean {
    return s1.equals(s2, ignoreCase = true)
}

/**
 * Return `""` if string equals null.
 *
 * @param s The string.
 * @return `""` if string equals null
 */
fun null2Length0(s: String?): String {
    return s ?: ""
}

/**
 * Return the length of string.
 *
 * @param s The string.
 * @return the length of string
 */
fun length(s: CharSequence?): Int {
    return s?.length ?: 0
}

/**
 * Set the first letter of string upper.
 *
 * @param s The string.
 * @return the string with first letter upper.
 */
fun upperFirstLetter(s: String?): String {
    if (s == null || s.isEmpty()) return ""
    return if (!Character.isLowerCase(s[0])) s else (s[0].toInt() - 32).toChar().toString() + s.substring(1)
}

/**
 * Set the first letter of string lower.
 *
 * @param s The string.
 * @return the string with first letter lower.
 */
fun lowerFirstLetter(s: String?): String {
    if (s == null || s.isEmpty()) return ""
    return if (!Character.isUpperCase(s[0])) s else (s[0].toInt() + 32).toChar().toString() + s.substring(1)
}

/**
 * Reverse the string.
 *
 * @param s The string.
 * @return the reverse string.
 */
fun reverse(s: String?): String {
    if (s == null) return ""
    val len = s.length
    if (len <= 1) return s
    val mid = len shr 1
    val chars = s.toCharArray()
    var c: Char
    for (i in 0 until mid) {
        c = chars[i]
        chars[i] = chars[len - i - 1]
        chars[len - i - 1] = c
    }
    return String(chars)
}

/**
 * Convert string to DBC.
 *
 * @param s The string.
 * @return the DBC string
 */
fun toDBC(s: String?): String {
    if (s == null || s.isEmpty()) return ""
    val chars = s.toCharArray()
    var i = 0
    val len = chars.size
    while (i < len) {
        when {
            chars[i].toInt() == 12288 -> chars[i] = ' '
            chars[i].toInt() in 65281..65374 -> chars[i] = (chars[i].toInt() - 65248).toChar()
            else -> chars[i] = chars[i]
        }
        i++
    }
    return String(chars)
}

/**
 * Convert string to SBC.
 *
 * @param s The string.
 * @return the SBC string
 */
fun toSBC(s: String?): String {
    if (s == null || s.isEmpty()) return ""
    val chars = s.toCharArray()
    var i = 0
    val len = chars.size
    while (i < len) {
        when {
            chars[i] == ' ' -> chars[i] = 12288.toChar()
            chars[i].toInt() in 33..126 -> chars[i] = (chars[i].toInt() + 65248).toChar()
            else -> chars[i] = chars[i]
        }
        i++
    }
    return String(chars)
}